addon/index.d.ts

```diff
@@ -0,0 +1,828 @@
+
+export type Evented = {
+  on(event: string, callback: (...args: any[]) => void): void;
+  off(event: string, callback: (...args: any[]) => void): void;
+} | {
+  one(event: string, callback: (...args: any[]) => void): void;
+} | {
+  addEventListener(event: string, callback: (...args: any[]) => void): void;
+  removeEventListener(event: string, callback: (...args: any[]) => void): void;
+};
+
+/**
+ * A Task is a cancelable, restartable, asynchronous operation that
+ * is driven by a generator function. Tasks are automatically canceled
+ * when the object they live on is destroyed (e.g. a Component
+ * is unrendered).
+ *
+ * To define a task, use the `task(...)` function, and pass in
+ * a generator function, which will be invoked when the task
+ * is performed. The reason generator functions are used is
+ * that they (like the proposed ES7 async-await syntax) can
+ * be used to elegantly express asynchronous, cancelable
+ * operations.
+ *
+ * You can also define an
+ * <a href="/#/docs/encapsulated-task">Encapsulated Task</a>
+ * by passing in an object that defined a `perform` generator
+ * function property.
+ *
+ * The following Component defines a task called `myTask` that,
+ * when performed, prints a message to the console, sleeps for 1 second,
+ * prints a final message to the console, and then completes.
+ *
+ * ```js
+ * import { task, timeout } from 'ember-concurrency';
+ * export default Component.extend({
+ *   myTask: task(function * () {
+ *     console.log("Pausing for a second...");
+ *     yield timeout(1000);
+ *     console.log("Done!");
+ *   })
+ * });
+ * ```
+ *
+ * ```hbs
+ * <button {{action myTask.perform}}>Perform Task</button>
+ * ```
+ *
+ * By default, tasks have no concurrency constraints
+ * (multiple instances of a task can be running at the same time)
+ * but much of a power of tasks lies in proper usage of Task Modifiers
+ * that you can apply to a task.
+ *
+ * @param taskFn The generator function backing the task.
+ */
+export function task<T extends TaskFunction<any, any[]>>(taskFn: T):
+  TaskProperty<TaskFunctionReturnType<T>, TaskFunctionArgs<T>>;
+
+/**
+ * "Task Groups" provide a means for applying
+ * task modifiers to groups of tasks. Once a {@linkcode Task} is declared
+ * as part of a group task, modifiers like `drop()` or `restartable()`
+ * will no longer affect the individual `Task`. Instead those
+ * modifiers can be applied to the entire group.
+ *
+ * ```js
+ * import { task, taskGroup } from 'ember-concurrency';
+ *
+ * export default Controller.extend({
+ *   chores: taskGroup().drop(),
+ *
+ *   mowLawn:       task(taskFn).group('chores'),
+ *   doDishes:      task(taskFn).group('chores'),
+ *   changeDiapers: task(taskFn).group('chores')
+ * });
+ * ```
+ *
+ * @param taskFn The generator function backing the task.
+ * @returns {TaskGroup}
+ */
+export function taskGroup<T extends TaskFunction<any, any[]>>(taskFn: T):
+  TaskGroup<TaskFunctionReturnType<T>>;
```

Comment:
I am pretty sure this is wrong:

1. it does not (should not) take a `taskFn`
2. it returns a `TaskGroupProperty` instead of a `TaskGroup`

Reply:
https://github.com/machty/ember-concurrency/pull/358

Reply:
#358 is merged, and this has been rebased to add docs and tests